lasagne/updates.py

```diff
@@ -567,6 +575,123 @@ def adam(loss_or_grads, params, learning_rate=0.001, beta1=0.9,
+        very small or zero norms.
+
+    Returns
+    -------
+    List of:
+
+        * List of scaled tensors
+        * Total norm
+        * Scaling multiplier
+
+    Notes
+    -----
+    The total norm and the scaling multiplier can be used to monitor training.
+
+    """
+    if not isinstance(tensor_vars, (list, tuple)):
+        tensor_vars = [tensor_vars]
+
+    sq = 0
+    for tensor in tensor_vars:
+        sq += (tensor**2).sum()
+
+    norm = T.sqrt(sq)
+    threshold = utils.floatX(threshold)
+    scale = threshold / (norm + epsilon)
+    multiplier = T.switch(norm < threshold, 1.0, scale)
+    tensor_vars_scaled = [step*multiplier for step in tensor_vars]
+
+    return tensor_vars_scaled, norm, multiplier
+
+
+def total_norm_constraint(tensor_or_tensor_vars, max_norm, epsilon=1e-7,
+                          return_norms=False):
+    """Rescales a list of tensors based on their combined norm
+
+    If the combined norm of the input tensors exceeds the threshold then all
+    tensors are rescaled such that the combined norm is equal to the threshold.
+
+    Scaling the norms of the gradients is often used when training recurrent
+    neural networks [1_].
+
+    Parameters
+    ----------
+    tensor_or_tensor_vars: TensorVariable or list of TensorVariables
+        Tensors to be rescaled.
+    threshold : float
+        Threshold value for total norm.
+    epsilon : scalar, optional
+        Value used to prevent numerical instability when dividing by
+        very small or zero norms.
+    return_norms : bool
+        If true the total norm is also returned
+
+    Returns
+    -------
+    List of:
+
+        * List of scaled tensors
+        * Total norm (optional with return_norms)
+
+    Usage
+    -----
+    >>> from lasagne.layers import InputLayer, DenseLayer
+    >>> import lasagne
+    >>> from lasagne.updates import sgd, total_norm_constraint
+    >>> x = T.matrix()
+    >>> y = T.ivector()
+    >>> l_in = InputLayer((5, 10))
+    >>> l1 = DenseLayer(l_in, num_units=7, nonlinearity=T.nnet.softmax)
+    >>> output = lasagne.layers.get_output(l1, x)
+    >>> cost = T.mean(T.nnet.categorical_crossentropy(output, y))
+    >>> all_params = lasagne.layers.get_all_params(l1)
+    >>> all_grads = T.grad(cost, all_params)
+    >>> scaled_grads = total_norm_constraint(all_grads, 5)
+    >>> updates = sgd(scaled_grads, all_params, learning_rate=0.1)
+
+    Notes
+    -----
+    The total norm can be used to monitor training.
+
+    References
+    ----------
+    .. [1] Sutskever, I., Vinyals, O., & Le, Q. V. (2014): Sequence to sequence
+       learning with neural networks. In Advances in Neural Information
+       Processing Systems (pp. 3104-3112).
+    """
+    if not isinstance(tensor_or_tensor_vars, (list, tuple)):
+        tensor_or_tensor_vars = [tensor_or_tensor_vars]
+
+    norm = T.sqrt(T.sum([T.sum(tensor**2) for tensor in
+                         tensor_or_tensor_vars]))
```

Comment:
It also passes the tests. I cant find the correct name for theano scalar so its called that unless you have better name?